pkg/controller/rolloutblock/rolloutblock_controller.go

```diff
@@ -0,0 +1,622 @@
+	recorder         record.EventRecorder
+
+	applicationLister  shipperlisters.ApplicationLister
+	applicationsSynced cache.InformerSynced
+
+	releaseLister  shipperlisters.ReleaseLister
+	releasesSynced cache.InformerSynced
+
+	rolloutBlockLister shipperlisters.RolloutBlockLister
+	rolloutBlockSynced cache.InformerSynced
+
+	rolloutblockWorkqueue workqueue.RateLimitingInterface
+}
+
+// NewController returns a new RolloutBlock controller.
+func NewController(
+	shipperClientset clientset.Interface,
+	informerFactory shipperinformers.SharedInformerFactory,
+	recorder record.EventRecorder,
+) *Controller {
+	applicationInformer := informerFactory.Shipper().V1alpha1().Applications()
+	releaseInformer := informerFactory.Shipper().V1alpha1().Releases()
+	rolloutblockInformer := informerFactory.Shipper().V1alpha1().RolloutBlocks()
+
+	glog.Info("Building a RolloutBlock controller")
+
+	controller := &Controller{
+		recorder:         recorder,
+		shipperClientset: shipperClientset,
+
+		applicationLister:  applicationInformer.Lister(),
+		applicationsSynced: applicationInformer.Informer().HasSynced,
+
+		releaseLister:  releaseInformer.Lister(),
+		releasesSynced: releaseInformer.Informer().HasSynced,
+
+		rolloutBlockLister: rolloutblockInformer.Lister(),
+		rolloutBlockSynced: rolloutblockInformer.Informer().HasSynced,
+
+		rolloutblockWorkqueue: workqueue.NewNamedRateLimitingQueue(
+			workqueue.DefaultControllerRateLimiter(),
+			"rolloutblock_controller_rolloutblocks",
+		),
+	}
+
+	glog.Info("Setting up event handlers")
+
+	releaseInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: controller.addRelease,
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				controller.updateRelease(oldObj, newObj)
+			},
+			DeleteFunc: controller.deleteRelease,
+		})
+
+	applicationInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: controller.addApplication,
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				controller.updateApplication(oldObj, newObj)
+			},
+			DeleteFunc: controller.deleteApplication,
+		})
+
+	rolloutblockInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: controller.addRolloutBlock,
+		})
+
+	return controller
+}
+
+func (c *Controller) addRelease(obj interface{}) {
+	rel, ok := obj.(*shipper.Release)
+	if !ok {
+		runtime.HandleError(fmt.Errorf("not a shipper.Release: %#v", obj))
+		return
+	}
+
+	key, err := cache.MetaNamespaceKeyFunc(rel)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	overrideRB, ok := rel.GetAnnotations()[shipper.RolloutBlocksOverrideAnnotation]
+	if !ok {
+		overrideRB = ""
+		return
+	}
+
+	overrideRBs := strings.Split(overrideRB, ",")
+	for _, rbKey := range overrideRBs {
+		rolloutBlockUpdater := RolloutBlockUpdater{
+			rbKey,
+			OverridingRelease,
+			key,
+			false,
+		}
+		c.rolloutblockWorkqueue.Add(rolloutBlockUpdater)
+	}
+}
+
+func (c *Controller) deleteRelease(obj interface{}) {
```

Comment:
Not sure the method name reflects on what's happening inside. Should it be rather called like: enqueueRolloutBlock(Updater)?

Reply:
I missed the initial intention of this method. I would suggest something like: onDeleteRelease instead, so it provider an event reactor context.